core/src/robot/LCPSolvers.cc

```diff
@@ -0,0 +1,221 @@
+                x[i] = clamp(x[i], - hiTmp, hiTmp);
+            }
+
+            // Check if still possible to terminate after complete update
+            if (checkAbs)
+            {
+                isSuccess = isSuccess && (std::abs(x[i] - xPrev) < tolAbs_);
+            }
+            if (checkRel && std::abs(x[i]) > EPS_DIVISION)
+            {
+                isSuccess = isSuccess && (std::abs((x[i] - xPrev) / x[i]) < tolRel_);
+            }
+        }
+
+        return isSuccess;
+    }
+
+    bool_t PGSSolver::ProjectedGaussSeidelSolver(matrixN_t & A,
+                                                 vectorN_t & b,
+                                                 vectorN_t const & lo,
+                                                 vectorN_t const & hi,
+                                                 std::vector<int32_t> const & fIdx,
+                                                 vectorN_t & x)
+    {
+        /* The implementation is partially adapted from Dart Simulator:
+           https://github.com/dartsim/dart/blob/master/dart/constraint/PgsBoxedLcpSolver.cpp */
+
+        // Single loop of standard Projected Gauss Seidel algorithm
+        bool_t isSuccess = ProjectedGaussSeidelIter(A, b, lo, hi, fIdx, true, false, x);
+
+        // If termination condition has been reached, return right now
+        if (isSuccess)
+        {
+            return true;
+        }
+
+        // Normalizing
+        for (int32_t i = 0; i < b.size(); ++i)
+        {
+            b[i] /= A(i, i);
+            A.row(i).array() /= A(i, i);
+        }
+
+        // Perform multiple PGS loop until convergence or max iter reached
+        for (uint32_t iter = 1; iter < maxIter_; ++iter)
+        {
+            isSuccess = ProjectedGaussSeidelIter(A, b, lo, hi, fIdx, false, true, x);
+            if (isSuccess)
+            {
+                return true;
+            }
+        }
+
+        // Impossible to converge
+        return false;
+    }
+
+    bool_t PGSSolver::BoxedForwardDynamics(pinocchio::Model const & model,
+                                           pinocchio::Data & data,
+                                           vectorN_t const & tau,
+                                           matrixN_t const & J,
+                                           vectorN_t const & gamma,
+                                           float64_t const & inv_damping,
+                                           vectorN_t const & lo,
+                                           vectorN_t const & hi,
+                                           std::vector<int32_t> const & fIdx)
+    {
+        // Define some proxies for convenience
+        matrixN_t & A = data.JMinvJt;
+        vectorN_t & f = data.lambda_c;
+
+        // Keep previous value if size did not change, reset it otherwise
+        if (f.size() != gamma.size())
+        {
+            f = vectorN_t::Zero(gamma.size());
+        }
+
+        // Compute the UDUt decomposition of data.M
```

Comment:
Please keep the copyright holders in all the files where you have done copy/pasting. This is a legal issue!

Reply:
Is it ok now ? https://github.com/Wandercraft/jiminy/blob/afd2a84af5a038e641ffd713af9a5e05bd7826ed/core/src/robot/LCPSolvers.cc#L140

Reply:
Thanks for accounting for this request.
I think it also does apply for DART if the code you mention is originally coming from this framework ;)

Reply:
In the particular case of DART, the rework is more significant. I use modern C++17 stl algorithms, Eigen-based linear algebra instead of custom operations on raw C flattened array pointers, and a small mathematical trick to speed-up computations by leveraging CPU vectorization. Both implementation are mathematically equivalent yet the implement is different (no copy-paste at all).

Reply:
If you say so, I won't check it. 
But as you're working for a company, it corresponds to some classic and standard legal issues I would say.

Reply:
What do you mean ? I should double check that I'm done violating any legal stuff ? Jiminy is not related to any company in practice, and I'm thinking to completely get rid of Wandercraft repository, which is basically a mirror of my own repository. Do you think it could help ?

Reply:
Let's discuss it on private channels for legal issues if you want. I let you contact me.